Description: Do not pass -m32/-m64 flags for native compilation
 
Author: Afif Elghraoui <afif@debian.org>
Bug-Debian: https://bugs.debian.org/828877
Forwarded: no
Last-Update: 2016-08-27
--- pkg-gridengine.orig/source/aimk
+++ pkg-gridengine/source/aimk
@@ -1411,17 +1411,7 @@
 #       set CC  = clanga
 #       set CFLAGS = "$CFLAGS -Wno-unused-value"
 
-   if ($BUILDARCH == LINUXPPC) then
-      set CFLAGS = "-m32 $CFLAGS" # fixme: why?
-      set LFLAGS = "-m32 $LFLAGS"
-      set SHARED_LFLAGS = "-m32 $SHARED_LFLAGS"
-   else if ($BUILDARCH == LINUXSPARC64 || $BUILDARCH == LINUXPPC64 || $BUILDARCH == LINUXMIPS64) then
-      if ($FORCE_32BIT != 1) then
-         set CFLAGS = "-m64 $CFLAGS"
-         set LFLAGS = "-m64 $LFLAGS"
-         set SHARED_LFLAGS = "-m64 $SHARED_LFLAGS"
-      endif
-   else if ($BUILDARCH == LINUXS390X && $FORCE_32BIT == 1) then
+   if ($BUILDARCH == LINUXS390X && $FORCE_32BIT == 1) then
       set CFLAGS = "-m31 $CFLAGS"
       set LFLAGS = "-m31 $LFLAGS"
       set SHARED_LFLAGS = "-m31 $SHARED_LFLAGS"
